### PR TITLE
[CSL-1798] Move definition of Block into cardano-sl-core

### PR DESCRIPTION
**Pull request by pva701**

### CODE REVIEW

File: core/Pos/Core/Txp.hs [hunk 1]
```diff
@@ -42,26 +43,27 @@ module Pos.Txp.Core.Types
        -- * Block message
```
Comment (gromakovsky) [inline core/Pos/Core/Txp.hs]: `-- * Block message`? 🤔 

I bet some changes will be requested eventually, so consider removing this line.

File: explorer/src/Pos/Binary/Explorer.hs [hunk 2]
```diff
@@ -6,9 +6,8 @@ import           Universum
 
 import           Pos.Binary              ()
 import           Pos.Binary.Class        (Cons (..), Field (..), deriveSimpleBi)
-import           Pos.Core                (HeaderHash, Timestamp)
+import           Pos.Core                (HeaderHash, Timestamp, TxUndo)
```
Comment (gromakovsky) [inline explorer/src/Pos/Binary/Explorer.hs]: I suppose `TxUndo` shouldn't be in core, but it's rather small, so I won't be too sad seeing it there.

File: core/Pos/Core/Delegation.hs [hunk 3]
```diff
@@ -0,0 +1,56 @@
+-- | Delegation payload types.
+
+module Pos.Core.Delegation
```
Comment (gromakovsky) [inline core/Pos/Core/Delegation.hs]: I believe this module should also contains some stuff from `Pos.Core.Types`, namely:
```
       , ProxySigLight
       , ProxySKLight
       , ProxySigHeavy
       , ProxySKHeavy
```

File: core/Pos/Core/Ssc.hs [hunk 4]
```diff
@@ -112,7 +110,7 @@ mkCommitmentsMapUnsafe = CommitmentsMap
 
 -- | Opening reveals secret.
 newtype Opening = Opening
-    { getOpening :: (AsBinary Secret)
+    { getOpening :: AsBinary Secret
```
Comment (gromakovsky) [inline core/Pos/Core/Ssc.hs]: palekhchi

File: lib/src/Pos/Binary/Slog.hs [hunk 5]
```diff
@@ -1,6 +1,6 @@
 -- | Binary serialization of Slog types.
 
-module Pos.Binary.Block.Slog
+module Pos.Binary.Slog
```
Comment (gromakovsky) [inline lib/src/Pos/Binary/Slog.hs]: Why?

Comment (gromakovsky) [reply lib/src/Pos/Binary/Slog.hs]: Well, maybe we should separate `Block` and `Slog`, but currently `Slog` is part of `Block`, so I don't see a reason why this module isn't called `Pos.Binary.Block.Slog` 🤷‍♂️

File: lib/src/Pos/Block/Base.hs [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+
+----------------------------------------------------------------------------
+-- Main smart constructors
+----------------------------------------------------------------------------
+
+type SanityConstraintM
+     = ( HasDifficulty BlockHeader
+       , HasHeaderHash BlockHeader
+       , HasConfiguration
+       )
```
Comment (gromakovsky) [inline lib/src/Pos/Block/Base.hs]: I think `HasDifficulty BlockHeader` and `HasHeaderHash BlockHeader` are in scope or can be added into scope by importing something from `Core`. So I guess `SanityConstraintM` can be deleted and `HasConfiguration` should be enough.

File: lib/src/Pos/Block/Base.hs [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+
+----------------------------------------------------------------------------
+-- Genesis smart constructors
+----------------------------------------------------------------------------
+
+type SanityConstraintG
+     = ( HasDifficulty BlockHeader
+       , HasHeaderHash BlockHeader
+       , HasConfiguration
+       )
```
Comment (gromakovsky) [inline lib/src/Pos/Block/Base.hs]: This one is the same as `SanityConstraintM`, I don't think it's a good idea to separate them.

File: ssc/Pos/Ssc/Class/LocalData.hs [hunk 8]
```diff
@@ -0,0 +1,46 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE RankNTypes          #-}
+
+-- | This module defines type class for local data storage.
+
+module Pos.Ssc.Class.LocalData
```
Comment (neongreen) [inline ssc/Pos/Ssc/Class/LocalData.hs]: Bad merge.

File: core/Pos/Core/Update.hs [hunk 9]
```diff
@@ -413,59 +408,3 @@ mkUpdateProof = hash
 -- TODO: I don't like that 'Some' is used here
 -- —@neongreen
 type UpdateBlock = Either (Some IsGenesisHeader) (Some IsMainHeader, UpdatePayload)
```
[the remark this answers is not in the document]
Comment (pva701) [reply core/Pos/Core/Update.hs]: how?

File: ssc/Pos/Ssc/Base.hs [hunk 10]
```diff
@@ -2,7 +2,7 @@
 
 -- | Core functions from SSC.
 
-module Pos.Ssc.Core.Functions
+module Pos.Ssc.Base
```
[the remark this answers is not in the document]
Comment (pva701) [reply ssc/Pos/Ssc/Base.hs]: `Slavka`

File: lib/src/Pos/Util/JsonLog.hs [hunk 11]
```diff
@@ -36,17 +36,18 @@ import           Mockable                      (Catch, Mockable, realTime)
...
 import           Pos.Communication.Relay.Logic (InvReqDataFlowLog)
 import           Pos.Core                      (HasConfiguration, SlotId (..), gbHeader,
                                                 gbhPrevBlock, getSlotIndex, headerHash,
                                                 mkLocalSlotIndex)
+import           Pos.Core.Block                (Block, mainBlockTxPayload)
+import           Pos.Core.Block.Genesis.Lens   (genBlockEpoch)
+import           Pos.Core.Block.Main.Lens      (mainBlockSlot)
+import           Pos.Core.Txp                  (txpTxs)
 import           Pos.Crypto                    (hash, hashHexF)
-import           Pos.Txp                       (JLTxR (..), MemPoolModifyReason, txpTxs)
```
[the remark this answers is not in the document]
Comment (pva701) [reply lib/src/Pos/Util/JsonLog.hs]: `import           Pos.Core.Txp                  (txpTxs)`

File: ssc/Pos/Ssc/Base.hs (hunk 10 above)
[the remark this answers is not in the document]
Comment (pva701) [reply ssc/Pos/Ssc/Base.hs]: Your feeling is right

File: core/Pos/Core/Delegation.hs (hunk 3 above)
Comment (pva701) [reply core/Pos/Core/Delegation.hs]: Somnevaus'

```
    Module imports form a cycle:
             module ‘Pos.Crypto.Signing.Tag’ (./Pos/Crypto/Signing/Tag.hs)
            imports ‘Pos.Core.Configuration.Protocol’ (./Pos/Core/Configuration/Protocol.hs)
      which imports ‘Pos.Core.Genesis.Types’ (./Pos/Core/Genesis/Types.hs)
      which imports ‘Pos.Core.Delegation’ (./Pos/Core/Delegation.hs)
      which imports ‘Pos.Core.Configuration’ (./Pos/Core/Configuration.hs)
      which imports ‘Pos.Core.Genesis.Canonical’ (./Pos/Core/Genesis/Canonical.hs)
      which imports ‘Pos.Core.Address’ (./Pos/Core/Address.hs)
      which imports ‘Pos.Crypto.Signing’ (./Pos/Crypto/Signing.hs)
      which imports ‘Pos.Crypto.Signing.Tag’ (./Pos/Crypto/Signing/Tag.hs)
```

File: core/Pos/Core/Update.hs (hunk 9 above)
[the remark this answers is not in the document]
Comment (neongreen) [reply core/Pos/Core/Update.hs]: By not using them

All functions that were taking a XxxBlock will now take a Block

File: core/Pos/Core/Delegation.hs (hunk 3 above)
Comment (gromakovsky) [reply core/Pos/Core/Delegation.hs]: But it seems you don't need full `HasConfiguration` here, I believe you need only a small part (so you can import only `Pos.Core.Configuration.SmallPart`) which might resolve the problem.

File: core/Pos/Core/Update.hs (hunk 9 above)
[the remark this answers is not in the document]
Comment (gromakovsky) [reply core/Pos/Core/Update.hs]: But then update system functions will have access to TxpPayload, for example, and we won't be sure that it doesn't use TxpPayload.

**Comment by gromakovsky** [comment]
I will approve after @neongreen approves.

**Pull request merged**